test/xds/xds_client_outlier_detection_test.go

```diff
@@ -0,0 +1,218 @@
+	}
+	return cluster
+}
+
+// TestOutlierDetectionWithOutlier tests the Outlier Detection Balancer e2e. It
+// spins up three backends, one which consistently errors, and configures the
+// ClientConn using xDS to connect to all three of those backends. The Outlier
+// Detection Balancer should eject the connection to the backend which
+// constantly errors, and thus RPC's should mainly go to backend 1 and 2.
+func (s) TestOutlierDetectionWithOutlier(t *testing.T) {
+	oldOD := envconfig.XDSOutlierDetection
+	envconfig.XDSOutlierDetection = true
+	internal.RegisterOutlierDetectionBalancerForTesting()
+	defer func() {
+		envconfig.XDSOutlierDetection = oldOD
+		internal.UnregisterOutlierDetectionBalancerForTesting()
+	}()
+
+	managementServer, nodeID, _, resolver, cleanup := e2e.SetupManagementServer(t, nil)
+	defer cleanup()
+
+	// counters for how many times backends got called
+	var count1, count2, count3 int
+
+	// Working backend 1.
+	port1, cleanup1 := startTestService(t, &stubserver.StubServer{
+		EmptyCallF: func(context.Context, *testpb.Empty) (*testpb.Empty, error) {
+			count1++
+			return &testpb.Empty{}, nil
+		},
+		Address: "localhost:0",
+	})
+	defer cleanup1()
+
+	// Working backend 2.
+	port2, cleanup2 := startTestService(t, &stubserver.StubServer{
+		EmptyCallF: func(context.Context, *testpb.Empty) (*testpb.Empty, error) {
+			count2++
+			return &testpb.Empty{}, nil
+		},
+		Address: "localhost:0",
+	})
+	defer cleanup2()
+	// Backend 3 that will always return an error and eventually ejected.
+	port3, cleanup3 := startTestService(t, &stubserver.StubServer{
+		EmptyCallF: func(context.Context, *testpb.Empty) (*testpb.Empty, error) {
+			count3++
+			return nil, errors.New("some error")
+		},
+		Address: "localhost:0",
+	})
+	defer cleanup3()
+
+	const serviceName = "my-service-client-side-xds"
+	resources := defaultClientResourcesMultipleBackendsAndOD(e2e.ResourceParams{
+		DialTarget: serviceName,
+		NodeID:     nodeID,
+		Host:       "localhost",
+		SecLevel:   e2e.SecurityLevelNone,
+	}, []uint32{port1, port2, port3})
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*100)
+	defer cancel()
+	if err := managementServer.Update(ctx, resources); err != nil {
+		t.Fatal(err)
+	}
+
+	cc, err := grpc.Dial(fmt.Sprintf("xds:///%s", serviceName), grpc.WithTransportCredentials(insecure.NewCredentials()), grpc.WithResolvers(resolver))
+	if err != nil {
+		t.Fatalf("failed to dial local test server: %v", err)
+	}
+	defer cc.Close()
+
+	client := testgrpc.NewTestServiceClient(cc)
+	for i := 0; i < 2000; i++ {
+		// Can either error or not depending on the backend called.
+		if _, err := client.EmptyCall(ctx, &testpb.Empty{}, grpc.WaitForReady(true)); err != nil && !strings.Contains(err.Error(), "some error") {
+			t.Fatalf("rpc EmptyCall() failed: %v", err)
+		}
+		time.Sleep(time.Millisecond)
+	}
+
+	// Backend 1 should've gotten more than 1/3rd of the load as backend 3
```

Comment:
I'm guessing the eventual expected behavior is that RPCs get roundrobin-ed across backends 1 and 2. Is it possible to check that instead of relying on the counts? Would that be cleaner? What do you think?

Reply:
Thanks for the suggestion. I like the idea. A few caveats here. The ejected SubConns causing only a round robin across 1 and 2 will be unejected in ejection analysis sweeps. Thus, this isn't the asymptotic behavior, but is behavior that should happen at a certain part in time. Thus, I can have a function similar to yours in the other PR I reviewed, where it waits over a time period to get a situation where you get only 1 and 2's in picks. Michael's interop test setup 5 backends and essentially made sure the ejected backend stopped receiving load and when unejected and healthy, continues to receive load. I would want further discussion on this though, since such a change could introduce flakiness and a lot of effort. I wouldn't mind changing this though, either to your suggestion or another one we come up with. @dfawley

Reply:
I'd be OK with adding a TODO here about possible options and coming back to it in a follow up PR. Thanks.